web3/utils/datatypes.py

```diff
@@ -0,0 +1,50 @@
+from functools import (
+    partial,
+)
+from eth_utils import (
+    to_dict,
+)
+from toolz.itertoolz import (
+    concat
+)
+
+
+def verify_attr(class_name, key, base):
+    if not hasattr(base, key):
+        raise AttributeError(
+            "Property {0} not found on {1} class. "
+            "`{1}.factory` only accepts keyword arguments which are "
+            "present on the {1} class".format(key, class_name)
+        )
+
+
+class PropertyCheckingFactory(type):
+    @to_dict
+    def normalize_properties(attributes, values, normalizers):
+        for attribute, value in zip(attributes, values):
+            if attribute in normalizers:
+                normalizer = normalizers[attribute]
+                yield attribute, normalizer(value)
+            else:
+                yield attribute, value
+
+    def __init__(cls, name, bases, namespace, **kargs):
+        # see PEP487.  To accept kwargs in __new__, they need to be
+        # filtered out here.
+        super().__init__(name, bases, namespace)
+
+    def __new__(mcs, name, bases, namespace, normalizers={}):
+        for key in namespace:
+            verify_key_attr = partial(verify_attr, name, key)
```

Comment:
Instead of using `partial` I'll recommend looking into the [`cytoolz.curry`](http://toolz.readthedocs.io/en/latest/api.html#toolz.functoolz.curry) which makes for a nicer partialing API